### PR TITLE
Support TIMESTAMP_SECONDS, TIMESTAMP_MILLIS and TIMESTAMP_MICROS [databricks]

### LINKED ISSUE #308: [FEA] Spark 3.1 adding support for  TIMESTAMP_SECONDS, TIMESTAMP_MILLIS and TIMESTAMP_MICROS functions
Spark 3.1  Adds TIMESTAMP_SECONDS, TIMESTAMP_MILLIS and TIMESTAMP_MICROS functions

see:
https://issues.apache.org/jira/browse/SPARK-31797

We should support when required

### PR DESCRIPTION
**Pull request by thirtiseven**
Closes #308 

This PR adds GPU support for functions TIMESTAMP_SECONDS, TIMESTAMP_MILLIS, and TIMESTAMP_MICROS, which are used to convert the number of seconds/milliseconds/microseconds from the Unix epoch to a timestamp.

Related PRs in Spark: [28534](https://github.com/apache/spark/pull/28534) and [28956](https://github.com/apache/spark/pull/28956).

### CODE REVIEW

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 1]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+  
+  override def dataType: DataType = TimestampType
+  override def outputTypeOverride: DType = DType.TIMESTAMP_MICROSECONDS
+
+  /**
+   * Test whether if input * multiplier will cause Long-overflow. In Math.multiplyExact, 
+   * if there is an integer-overflow, then it will throw an ArithmeticException "long overflow"
+   */
+  def checkLongMultiplicationOverflow(input: ColumnVector, multiplier: Long): Unit = {
+      withResource(input.max()) { maxValue =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:404]: `maxValue` is a scalar, you need to check if it is valid before reading its value.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 2]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+        Math.multiplyExact(maxValue.getLong, multiplier)
+      }
+      withResource(input.min()) { minValue =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:407]: `minValue` is a scalar, you need to check if it is valid before reading its value.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 3]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+                      inputTimesMicrosCv.castTo(DType.TIMESTAMP_MICROSECONDS)
+                  }
+                }
+            }
+          }
+        }
+      }
+    case dt: DecimalType =>
+      (input: GpuColumnVector) => {
+        println("DecimalType " + dt.precision + " " + dt.scale)
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:452]: Remove this debugging log

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 4]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+        val decimal128Type = DType.create(DType.DTypeEnum.DECIMAL128, -6)
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:453]: Why -6 is used for the scale here ?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 5]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+        val isEqual = withResource(input.getBase.castTo(decimal128Type)) { casted =>
+          withResource(casted.castTo(decimal128TypeAllScale)) { castedAllScale =>
+            withResource(input.getBase.castTo(decimal128TypeAllScale)) { original =>
+              original.equalTo(castedAllScale)
+            }
+          }
+        }
+        val roundUnnecessity = withResource(isEqual) { isEqual =>
+          withResource(isEqual.all()) { all =>
+            all.getBoolean()
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:466]: You need to check if all is valid before getting its value.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 6]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+          }
+        }
+        if (!roundUnnecessity) {
+          throw new ArithmeticException("Rounding necessary")
+        }
+        val mul = withResource(input.getBase.castTo(decimal128Type)) { decimal128 =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:472]: Better to comment why this casting is needed.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 7]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+          mul.close()
+          throw new java.lang.ArithmeticException("Overflow")
+        }
+        withResource(mul) { mul =>
+          withResource(mul.castTo(DType.INT64)) { longs =>
+            longs.asTimestampMicroseconds()
+          }
+        }
+      }
+    case _ =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:506]: List the possible remaining types here. And use this `_` for unmatched types with throwing an exception saying "unsupported types $type"

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 8]
```diff
@@ -390,6 +391,171 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+      (input: GpuColumnVector) =>
+        withResource(input.getBase.castTo(DType.INT64)) { longs =>
+          // Not possible to overflow for Int, Short and Byte
+          longs.asTimestampSeconds()
+        }
+}
+
+  override def doColumnar(input: GpuColumnVector): ColumnVector = {
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:514]: This implementation can be moved into the parent trait  `GpuNumberToTimestampUnaryExpression`, then you don't need to duplicate it three times in these 3 children expressions.

**Comment by thirtiseven** [comment]
@firestarman Thanks for review! I think I have addressed all your comments.

**Comment by thirtiseven** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 9]
```diff
@@ -390,6 +391,182 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+                      inputTimesMicrosCv.castTo(DType.TIMESTAMP_MICROSECONDS)
+                  }
+                }
+            }
+          }
+        }
+      }
+    case dt: DecimalType =>
+      (input: GpuColumnVector) => {
+        // SecondsToTimestamp only supports decimals with a scale of 6 or less, which can be
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:462]: This is confusing. It indicates that Spark's SecondsToTimestamp does not support decimals with a scaled of 6 or less.  But it does. You even test it.

Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala]: SecondsToTimestamp does support decimals with a scale of 6 or less, which can be represented as microseconds. An exception will be thrown if the scale is more than 6. I updated the comment here.

File: integration_tests/src/main/python/date_time_test.py [hunk 10]
```diff
@@ -438,3 +438,56 @@ def do_join_cast(spark):
...
     assert_gpu_and_cpu_are_equal_collect(do_join_cast)
+
+# A None is appended here to make sure the output type is nullable, otherwise the test will fail
+# because inf and nan will be converted to null. It is workaround to avoid triggering an optimization 
+# in Spark row processing based on codegen, which is not supported by the plugin.
+ts_float_gen = SetValuesGen(FloatType(), [0.0, -0.0, 1.0, -1.0, 1.234567, -1.234567, float('inf'), float('-inf'), float('nan'), None])
+# (-62135510400, 253402214400) is the range of seconds that can be represented by timestamp_seconds 
+# considering the influence of time zone.
+seconds_gens = [LongGen(min_val=-62135510400, max_val=253402214400), IntegerGen(), ShortGen(), ByteGen(), 
+                DoubleGen(min_exp=0, max_exp=32), ts_float_gen, DecimalGen(16, 6), DecimalGen(13, 3), DecimalGen(10, 0), DecimalGen(6, 6)]
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/date_time_test.py]: Done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 11]
```diff
@@ -390,6 +391,182 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+        val mul = withResource(Scalar.fromLong(DateTimeConstants.MICROS_PER_SECOND)) { scalar =>
+          input.getBase.mul(scalar)
+        }
+        withResource(mul) { mul =>
+          mul.asTimestampMicroseconds()
+        }
+      }
+    case DoubleType | FloatType =>
+      (input: GpuColumnVector) => {
+        // basicly copied from GpuCast
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala]: Yes, but [spark#31831](https://github.com/apache/spark/pull/31831) added the upcast at the same version as the hasCastFloatTimestampUpcast from false to true, so GpuCast can be called here safely. I didn't notice that, updated.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 12]
```diff
@@ -390,6 +391,182 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
   }
 }
 
+trait GpuNumberToTimestampUnaryExpression extends GpuUnaryExpression {
+  
+  override def dataType: DataType = TimestampType
+  override def outputTypeOverride: DType = DType.TIMESTAMP_MICROSECONDS
+
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:398]: Thanks! Done.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala (hunk 4 above)
Comment (thirtiseven) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala]: SecondsToTimestamp does support decimals with a scale of 6 or less, which can be represented as microseconds. An exception will be thrown if the scale is more than 6. I updated the comment here.

**Comment by thirtiseven** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 13]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+}
+
+case class GpuSecondsToTimestamp(child: Expression) extends GpuNumberToTimestampUnaryExpression {
+
+  override def nullable: Boolean = child.dataType match {
+    case _: FloatType | _: DoubleType => true
+    case _ => child.nullable
+  }
+
+  protected lazy val convertTo: GpuColumnVector => ColumnVector = child.dataType match {
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:430]: We can also add a `@transient` for this `covnertTo` just like [Spark does](https://github.com/apache/spark/blob/v3.1.1/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L496C3-L496C13).

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 14]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+    case dt: DecimalType =>
+      (input: GpuColumnVector) => {
+        // SecondsToTimestamp supports decimals with a scale of 6 or less, which can be represented
+        // as microseconds. An exception will be thrown if the scale is more than 6.
+        val decimalTypeSupported = DType.create(DType.DTypeEnum.DECIMAL128, -6)
+        if (dt.scale > 6) {
+          // Match the behavior of `BigDecimal.longValueExact()`, if decimal is equal to the value
+          // casted to decimal128 with scale 6, then no rounding is necessary.
+          val decimalTypeAllScale = DType.create(DType.DTypeEnum.DECIMAL128, -dt.scale)
+          val isEqual = withResource(input.getBase.castTo(decimalTypeSupported)) { casted =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:454]: `casted` can be closed rigth away when getting `castedAllScale`.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 15]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+    case DoubleType | FloatType =>
+      (input: GpuColumnVector) => {
+        GpuCast.doCast(input.getBase, input.dataType, TimestampType, false, false, false)
+      }
+    case dt: DecimalType =>
+      (input: GpuColumnVector) => {
+        // SecondsToTimestamp supports decimals with a scale of 6 or less, which can be represented
+        // as microseconds. An exception will be thrown if the scale is more than 6.
+        val decimalTypeSupported = DType.create(DType.DTypeEnum.DECIMAL128, -6)
+        if (dt.scale > 6) {
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:450]: NIT:
We can put this rounding check in a separate method, just like `checkLongMultiplicationOverflow`.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 16]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+            decimal.mul(scalar, decimalTypeSupported)
+          }
+        }
+        // Match the behavior of `BigDecimal.longValueExact()`:
+        val greaterThanMax = withResource(Scalar.fromLong(Long.MaxValue)) { longMax =>
+            mul.greaterThan(longMax)
+        }
+        val largerThanLongMax = withResource(greaterThanMax) { greaterThanMax =>
+          withResource(greaterThanMax.any()) { any =>
+            any.getBoolean()
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:482]: Need to check the `any`'s validity.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 17]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+          }
+        }
+        lazy val smallerThanLongMin: Boolean = {
+          val lessThanMin = withResource(Scalar.fromLong(Long.MinValue)) { longMin =>
+              mul.lessThan(longMin)
+          }
+          withResource(lessThanMin) { lessThanMin =>
+            withResource(lessThanMin.any()) { any =>
+              any.getBoolean()
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:491]: Need to check the any's validity.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 18]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+            throw new ArithmeticException("Rounding necessary")
+          }
+        }
+        // Cast to decimal128 to avoid overflow
+        val mul = withResource(input.getBase.castTo(decimalTypeSupported)) { decimal =>
+          withResource(Scalar.fromLong(DateTimeConstants.MICROS_PER_SECOND)) { scalar =>
+            decimal.mul(scalar, decimalTypeSupported)
+          }
+        }
+        // Match the behavior of `BigDecimal.longValueExact()`:
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:476]: You need to protect the `mul` when computing the `largerThanLongMax` and `smallerThanLongMin`. You can do it like
```
closeOnExcept(mul) { _ =>
   val greaterThanMax = ....
   val smallerThanLongMin = ....
   if (largerThanLongMax || smallerThanLongMin) {
       throw new java.lang.ArithmeticException("Overflow")
    }
}
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 19]
```diff
@@ -390,6 +391,167 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+            withResource(lessThanMin.any()) { any =>
+              any.getBoolean()
+            }
+          }
+        }
+        if (largerThanLongMax || smallerThanLongMin) {
+          mul.close()
+          throw new java.lang.ArithmeticException("Overflow")
+        }
+        withResource(mul) { mul =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:499]: NIT:
```suggestion
        withResource(mul) { _ =>
```

File: integration_tests/src/main/python/date_time_test.py [hunk 20]
```diff
@@ -438,3 +438,53 @@ def do_join_cast(spark):
...
+        error_message='long overflow')
+    
+@pytest.mark.parametrize('data_gen', [DecimalGen(7, 7), DecimalGen(20, 7)], ids=idfn)
+def test_timestamp_seconds_rounding_necessary(data_gen):
+    assert_gpu_and_cpu_error(
+        lambda spark : unary_op_df(spark, data_gen).selectExpr("timestamp_seconds(a)").collect(),
+        conf={},
+        error_message='Rounding necessary')
+    
+@pytest.mark.parametrize('data_gen', [DecimalGen(19, 6), DecimalGen(20, 6)], ids=idfn)
```
[the remark this answers is not in the document]
Comment (thirtiseven) [reply integration_tests/src/main/python/date_time_test.py:466]: Check for rounding will be prior to check for overflow, which is what I want to test with the Decimal(20,7) case.

**Comment by thirtiseven** [comment]
@firestarman Thanks for the review. All done, please take another look.

**Comment by revans2** [comment]
build

**Review by revans2 — APPROVED** [review]

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala [hunk 21]
```diff
@@ -390,6 +391,175 @@ abstract class UnixTimeExprMeta[A <: BinaryExpression with TimeZoneAwareExpressi
...
+              withResource(lessThanMin.any()) { any =>
+                any.isValid && any.getBoolean()
+              }
+            }
+          }
+          if (largerThanLongMax || smallerThanLongMin) {
+            throw new java.lang.ArithmeticException("Overflow")
+          }
+        }
+        withResource(mul) { _ =>
```
Comment (firestarman) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/datetimeExpressions.scala:507]: NIT: `mul` can be released just after getting `longs`
```
val longs = withResource(mul)(_.castTo(DType.INT64))
withResource(longs)(_.asTimestampMicroseconds())
```

**Comment by firestarman** [comment]
Looks good to me, only one nit

**Review by firestarman — APPROVED** [review]

**Comment by thirtiseven** [comment]
build

**Review by firestarman — APPROVED** [review]

**Comment by thirtiseven** [comment]
build

**Pull request merged**